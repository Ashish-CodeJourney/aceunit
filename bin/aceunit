#!/bin/bash

tool=objdump
name=fixtures

usage() {
    echo "Usage: $0 [-h] [-p PATTERN] [-t TOOL] OBJECT_FILE..." 1>&2
}

version() {
    echo "$0 version 1.0"
}

setType() {
    case "$1" in
    objdump) tool=objdump ;;
    nm) tool=nm ;;
    readelf) tool=readelf ;;
    *) echo "$0: error: unsupported symbol table tool '$1'" >&2 ; exit 1 ;;
    esac
}

help() {
    usage
    cat <<END
Generate AceUnit fixture list from compiled testcase object files.
The generated test suite is written to STDOUT.

Options:
  -b BINARY   Use BINARY instead of the default for the selected tool.
              Example: -b m68k-amigaos-objdump
  -h          Display this help text and stop.
  -n NAME     Use NAME instead of fixtures for the fixture table.
  -p PATTERN  Prefix test functions with the specified regex PATTERN;
              PATTERN needs to be understood by grep.
              Example: -p '[^_]\+_' to allow any pattern prefix followed by an underscore.
  -s PATTERN  Strip this prefix PATTERN from symbols before using them;
              PATTERN needs to be understood by grep.
              Example: -s _ to strip a leading underscore from the symbol names.
  -t TOOL     Use TOOL for displaying the symbol table.
              Defaults to objdump.
              Built-in alternatives: nm, readelf.
  -v          Print version information and stop.

Examples:
  $0 leapyear_test.o >testcases.c
    Scans leapyear_test.o for test functions and generates the test fixtures in testcases.c.

  $0 -t nm -b m68k-amigaos-nm -s _ leapyear_test.o >testcases.c
    Scans leapyear_test.o for test functions and generates the test fixtures in testcase.c.
    Uses nm to get the symbol table, with m68k-amigaos-nm as the binary of nm, and stripping leading underscores from the symbol names.
END
    exit 0
}

function getSymbols() {
    if ! which "$binary" >/dev/null ; then
        echo "$binary: File not found." 1>&2
        exit 2
    fi
    case "${tool}" in
    objdump) $binary -t "$1" | grep -E ' g     [F ] .text[	 ]+' | sed -e 's/.* '$strip'//' | grep '^\('$2$3'\)' ;;
    nm)      $binary    "$1" | grep -E ' T '                       | sed -e 's/.* '$strip'//' | grep '^\('$2$3'\)' ;;
    readelf) $binary -s "$1" | grep -E ' FUNC    GLOBAL DEFAULT '  | sed -e 's/.* '$strip'//' | grep '^\('$2$3'\)' ;;
    *) echo "$0: internal error: Unsupported tool ${tool}" 1>&2 ; exit 2 ;;
    esac
}

prefix=

while getopts "b:hn:p:s:t:v" o; do
    case "${o}" in
    b) binary="$OPTARG" ;;
    h) help ;;
    n) name="$OPTARG" ;;
    p) prefix="$OPTARG" ;;
    s) strip="$OPTARG" ;;
    t) setType "$OPTARG" ;;
    v) version; exit 0 ;;
    *) usage; exit 1 ;;
    esac
done

shift $((OPTIND - 1))
fixtures=$@
declare -A beforeAll afterAll beforeEach afterEach testcases

if [[ -z "$binary" ]] ; then
    binary=$tool
fi

for file in $fixtures; do
    basename=${file%.o}
     beforeAll[$basename]=$(getSymbols $file $prefix beforeAll)
      afterAll[$basename]=$(getSymbols $file $prefix afterAll)
    beforeEach[$basename]=$(getSymbols $file $prefix beforeEach)
     afterEach[$basename]=$(getSymbols $file $prefix afterEach)
     testcases[$basename]=$(getSymbols $file $prefix test)
done

cat << END
#include <aceunit.h>

END
for file in $fixtures; do
    basename=${file%.o}
    if [ -z "${testcases[$basename]}" ]; then echo "$@: warning: No test cases found." 1>&2; fi
    for extern in ${beforeAll[$basename]} ${afterAll[$basename]} ${beforeEach[$basename]} ${afterEach[$basename]} ${testcases[$basename]}; do echo "extern void $extern(void);"; done
done
for file in $fixtures; do
    basename=${file%.o}
    echo
    echo "void (*const testcases_$basename[])(void) = {"
    for testcase in ${testcases[$basename]}; do echo "    $testcase,"; done
    cat << END
    NULL,
};

const AceUnit_Fixture_t fixture_$basename = {
    ${beforeAll[$basename]:-NULL},
    ${afterAll[$basename]:-NULL},
    ${beforeEach[$basename]:-NULL},
    ${afterEach[$basename]:-NULL},
    testcases_$basename,
};

END
done
cat <<END
const AceUnit_Fixture_t *$name[] = {
END
for file in $fixtures; do
    basename=${file%.o}
    echo "    &fixture_$basename,"
done
cat <<END
    NULL,
};
END
