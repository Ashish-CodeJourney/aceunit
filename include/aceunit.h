#ifndef ACEUNIT_H
#define ACEUNIT_H

#include <stddef.h>
#include <stdbool.h>

/** Fixture information.
 * This will usually be auto-generated by `gentest.sh`.
 */
typedef struct {
    /** Pointer to the `beforeAll()` method of the fixture, if present, otherwise `NULL`.
     * `beforeAll()` will be called once per fixture, before anything else.
     */
    void (*const beforeAll)();

    /** Pointer to the `afterAll()` method of the fixture, if present, otherwise `NULL`.
     * `afterAll()` will be called once per fixture, after anything else.
     */
    void (*const afterAll)();

    /** Pointer to the `beforeEach()` method of the fixture, if present, otherwise `NULL`.
     * `beforeEach()` will be called once per testcase, before the testcase.
     */
    void (*const beforeEach)();

    /** Pointer to the `afterEach()` method of the fixture, if present, otherwise `NULL`.
     * `afterEach()` will be called once per testcase, after the testcase.
     */
    void (*const afterEach)();

    /** Pointer to the array of test cases. */
    void (*const *testcases)();
} AceUnit_Fixture_t;

/** Result of a test run. */
typedef struct {
    /** The number of test cases that were executed. */
    int testCaseCount;

    /** The number of test cases that completed successfully. */
    int successCount;

    /** The number of failures in testcases and fixture methods. */
    int failureCount;
} AceUnit_Result_t;

/** The global fixtures. */
extern const AceUnit_Fixture_t *fixtures[];

/** Run AceUnit for the given list of fixtures.
 *
 * @note Don't forget to initialize \p results.
 * The function does not initialize \p results intentionally.
 * This way, multiple runs can be collected together in the same results.
 *
 * @param fixtures Pointer to the fixtures to run, must not be `NULL`, must be terminated with `NULL`.
 * @param result Where to store the results, must not be `NULL`.
 */
extern void AceUnit_run(const AceUnit_Fixture_t *fixtures[], AceUnit_Result_t *result);

/** Run a function, catching some types of errors.
 * Implementations must guarantee that they only ever return `true` upon successful completion of the function.
 * Implementations are allowed to not return at all in case of errors.
 * What types of errors are caught and converted to `false` is implementation defined.
 * @param code Function to run.
 * @return Whether the function was executed successfully.
 * @retval true when the function was executed successfully.
 * @retval false otherwise
 */
extern bool runCatching(void(*code)());

/** Fail and abort a test case.
 * Implementations may use different mechanisms for how to actually fail the test case.
 * Popular implementations could be:
 * - `longjmp()` which would work anywhere including most freestanding environments.
 * - `abort()` which would work on all hosted environments.
 * - `exit()` which would work with `AceUnit_Fork`.
 */
extern void AceUnit_fail();

#ifndef assert
#include <stdio.h>
/** Assert a condition.
 * If the condition is \c false, this will call #AceUnit_fail().
 * @param cond  Condition to assert.
 */
#define assert(cond) \
    do { \
        if (!(cond)) { \
            fprintf(stderr, "%s:%d: %s: Assertion `%s' failed.\n", __FILE__, __LINE__, __func__, #cond); \
            AceUnit_fail(); \
        } \
    } while (false)
#endif

#endif
