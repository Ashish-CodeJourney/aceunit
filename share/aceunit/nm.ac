tools[nm]=nm

getSymbols_nm() {
    $binary    "$1" | grep -E ' T '                         | sed -e 's/.* '$strip'//' | grep '^\('$2'\)'
}
