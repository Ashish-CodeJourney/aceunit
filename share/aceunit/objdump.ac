tools[objdump]=objdump

getSymbols_objdump() {
    $binary -t "$1" | grep -E ' g     [F ] [^	 ]+[	 ]+' | sed -e 's/.* '$strip'//' | grep '^\('$2'\)'
}
