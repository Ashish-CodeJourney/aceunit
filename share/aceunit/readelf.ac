tools[readelf]=readelf

getSymbols_readelf() {
    $binary -s "$1" | grep -E ' FUNC    GLOBAL DEFAULT '    | sed -e 's/.* '$strip'//' | grep '^\('$2'\)'
}
