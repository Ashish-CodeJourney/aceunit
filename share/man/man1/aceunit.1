.TH ACEUNIT "1" "March 2023" "AceUnit 3.0" "User Commands"
.SH NAME
aceunit \- generate test fixtures from object files
.SH SYNOPSIS
.B aceunit
[\fI\,OPTION\/\fR]... [\fI\,OBJECT_FILE\/\fR]...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Scans object files for test cases and generates the fixtures code on stdout.
.SH OPTIONS
.TP
\fB\-h\fR
Display help text.
.TP
\fB\-p\fR \fI\,PATTERN\/\fR
Use PATTERN as prefix for finding test functions.
.SH DEPENDENCIES
\fBaceunit\fR requires \fBobjdump\fR, \fBreadelf\fR, or \fBnm\fR to work.
.SH EXAMPLES
.TP
\fBaceunit \-p '[^_]\+_' *_test.o >aceunit.c\fR
will scan all object files matching the pattern \fB*_test.o\fR and write the found fixtures to \fBaceunit.c\fR.
.SH AUTHOR
Written by Christian Hujer.
.SH "REPORTING BUGS"
AceUnit online help: <https://nelkinda.com/products/aceunit/>
.SH COPYRIGHT
Copyright \(co 2007, 2023 Christian Hujer
License BSD
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
\&\fBobjdump\fR\|(1),
\&\fBreadelf\fR\|(1),
\&\fBnm\fR\|(1),
\&\fBabort\fR\|(3),
\&\fBassert\fR\|(3),
\&\fBlongjmp\fR\|(3),
\&\fBsetjmp\fR\|(3),
\&\fBfork\fR\|(2),
\&\fBsignal\fR\|(2),
Full documentation: <https://nelkinda.com/products/aceunit/>
