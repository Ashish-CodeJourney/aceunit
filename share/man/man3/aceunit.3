.TH ACEUNIT "3" "March 2023" "AceUnit 3.0" "Libraries"
.SH NAME
aceunit, assert, AceUnit_runFixtures, AceUnit_fail \- run unit tests with AceUnit
.SH SYNOPSIS
.nf
.B #include <aceunit.h>
.PP
.BI "void AceUnit_fail();
.BI "void assert(scalar " expression ");
.PP
.BI "void AceUnit_run(const AceUnit_Fixture *" fixtures[] ", AceUnit_Result_t *" result ");
.BI "bool runCatching(void(*" code ")());
.BI "int main(int " argc ", const char *" argv[] ");
.fi
.SH DESCRIPTION
The functions described on this page are used to execute unit tests.
The
.BR AceUnit_fail()
function can be called from a unit test to abort the current test case and mark it failed.
The
.BR assert():
macro effectively does the same if its scalar expression evaluates to zero.
.PP
The
.BR AceUnit_run()
function can be called to run fixtures.
See the aceunit(1) program for how to generate fixtures.
The
.BR runCatching()
function is called by
.BR AceUnit_run()
to wrap a test function, possibly catching signals.
The
.BR main()
function is provided for convenience to execute the test cases.
.SH EXAMPLES
Consider the following module, which provides a function to see whether a year is a leap year:
.PP
.EX
// File: leapyear.h
#ifndef LEAPYEAR_H
#define LEAPYEAR_H

#include <stdbool.h>

extern bool isLeap(int year);

#endif

// File: leapyear.c
#include "leapyear.h"

bool isLeap(int year)
{
    return year % 4 == 0 && year % 100 != 0 || year % 400 == 0;
}
.EE
.PP
Then a unit test module could look like this:
.EX
// File: leapyear_test.o
#include <aceunit.h>

void testLeapYears() {
    assert(isLeap(0));
    assert(isLeap(4));
    assert(isLeap(400));
}

void testNonLeapYears() {
    assert(!isLeap(1));
    assert(!isLeap(100));
}
.EE
.PP
AceUnit could be used with the following command sequence:
.EX
cc -c -o leapyear.o leapyear.c
cc -c -o leapyear_test.o leapyear_test.c
aceunit leapyear_test.o >unit_test.c
cc -c -o unit_test.o unit_test.c
cc -o unit_test unit_test.o leapyear_test.o leapyear.o -laceunit
\./aceunit
.EE
.SH AUTHOR
Written by Christian Hujer.
.SH "REPORTING BUGS"
AceUnit online help: <https://nelkinda.com/products/aceunit/>
.SH COPYRIGHT
Copyright \(co 2007, 2023 Christian Hujer
License BSD
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
\&\fBaceunit\fR\|(1),
\&\fBabort\fR\|(3),
\&\fBassert\fR\|(3),
\&\fBlongjmp\fR\|(3),
\&\fBsetjmp\fR\|(3),
\&\fBfork\fR\|(2),
\&\fBsignal\fR\|(2),
Full documentation: <https://nelkinda.com/products/aceunit/>
