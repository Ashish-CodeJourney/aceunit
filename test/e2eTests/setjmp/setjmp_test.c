#include <stdbool.h>
#include <stdio.h>
#include <aceunit.h>

#define printme() do { fprintf(stderr, "%s:%d: %s\n", __FILE__, __LINE__, __func__); } while (0)

void test1() {
    printme();
}

void test2() {
    printme();
    assert(false);
}

void test3() {
    printme();
}
