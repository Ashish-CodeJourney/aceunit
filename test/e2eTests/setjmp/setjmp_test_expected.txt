setjmp_test.c:8: test1
setjmp_test.c:12: test2
setjmp_test.c:13: test2: Assertion `false' failed.
setjmp_test.c:17: test3
AceUnit: 3 test cases, 2 successful, 1 failed.
